source/dofs/dof_handler_policy.cc

```diff
@@ -1068,14 +1068,40 @@ namespace internal
       distribute_dofs (DoFHandler<dim,spacedim> &dof_handler,
                        NumberCache &number_cache) const
       {
+        // If the underlying shared::Tria allows artifical cells, we need to do
+        // some tricks here to make Sequential algorithms play nicely.
+        // Namely, we first restore original partition (without artificial cells)
+        // and then turn artificial cells on at the end of this function.
+        const parallel::shared::Triangulation<dim, spacedim> *tr =
+          (dynamic_cast<const parallel::shared::Triangulation<dim, spacedim>*> (&dof_handler.get_tria ()));
+        Assert(tr != 0, ExcInternalError());
+        typename parallel::shared::Triangulation<dim,spacedim>::active_cell_iterator
+        cell = dof_handler.get_tria().begin_active(),
+        endc = dof_handler.get_tria().end();
+        if (tr->get_subdomain_id_without_artificial().size()>0)
+          for (unsigned int index=0; cell != endc; cell++, index++)
+            cell->set_subdomain_id(tr->get_subdomain_id_without_artificial()[index]);
```

Comment:
this is where you would store the old subdomain into a local vector first, before setting it to a new value.